pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterworkspace
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	tenancyinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	tenancylister "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/reconciler/tenancy/clusterworkspace/deletion"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/metadata"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+func NewGCController(
+	kcpClient kcpclient.ClusterInterface,
+	metadataClient metadata.Interface,
+	workspaceInformer tenancyinformer.ClusterWorkspaceInformer,
+	discoverResourcesFn func(clusterName logicalcluster.LogicalCluster) ([]*metav1.APIResourceList, error),
+) *GCController {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "workspace-gc")
+
+	c := &GCController{
+		queue:           queue,
+		kcpClient:       kcpClient,
+		workspaceLister: workspaceInformer.Lister(),
+	}
+
+	workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+
+	c.deleter = deletion.NewWorkspacedResourcesDeleter(kcpClient, metadataClient, discoverResourcesFn, "ws-finalize")
+	c.workspaceSynced = workspaceInformer.Informer().HasSynced
+
+	return c
+}
+
+type GCController struct {
```

Comment:
this is not really GC, but just namespace and object creation, right? Would just call it `clusterworkspacedeletion` as a package. 

Also, don't mix it with the existing clusterworkspace controller. Deletion is separate IMO.

Reply:
renamed